# Order Reviews: Data Ingestion & Quality Checks

This notebook loads the Olist `order_reviews` CSV with Pandas, validates its structure and key behavior, and writes the validated raw records to the existing MySQL `order_reviews` table.

### Workflow

`CSV → Pandas inspection → data quality checks → grain/key investigation → MySQL load → validation`

Raw review records are preserved. Data-quality findings are used to decide how the table should be interpreted in later analysis rather than automatically deleting unusual records.

## 1. Load the CSV

The source file is stored under `data/csv`. `pd.read_csv()` reads it into a Pandas DataFrame called `reviews`.

In [ ]:
import pandas as pd
from pathlib import Path

DATA_PATH = Path("../data/csv")

reviews = pd.read_csv(DATA_PATH / "olist_order_reviews_dataset.csv")

## 2. Inspect the dataset

The first checks confirm the dataset dimensions, column names, sample rows, data types, and non-null counts.

In [4]:
reviews.shape

(99224, 7)

In [5]:
reviews.head()

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53


In [6]:
reviews.columns

Index(['review_id', 'order_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp'],
      dtype='str')

In [7]:
reviews.info()

<class 'pandas.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype
---  ------                   --------------  -----
 0   review_id                99224 non-null  str  
 1   order_id                 99224 non-null  str  
 2   review_score             99224 non-null  int64
 3   review_comment_title     11568 non-null  str  
 4   review_comment_message   40977 non-null  str  
 5   review_creation_date     99224 non-null  str  
 6   review_answer_timestamp  99224 non-null  str  
dtypes: int64(1), str(6)
memory usage: 5.3 MB


### Initial findings

- **99,224 rows** and **7 columns**
- `review_id`, `order_id`, and `review_score` are fully populated
- `review_comment_title` and `review_comment_message` contain many null values
- Date columns were initially read as strings and need conversion before date-based analysis

## 3. Convert date columns

`pd.to_datetime()` converts the two date fields from strings to Pandas datetime values so they can be sorted, compared, and used in duration calculations.

In [8]:
reviews["review_creation_date"] = pd.to_datetime(
    reviews["review_creation_date"]
)

reviews["review_answer_timestamp"] = pd.to_datetime(
    reviews["review_answer_timestamp"]
)

In [9]:
reviews.info()

<class 'pandas.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   review_id                99224 non-null  str           
 1   order_id                 99224 non-null  str           
 2   review_score             99224 non-null  int64         
 3   review_comment_title     11568 non-null  str           
 4   review_comment_message   40977 non-null  str           
 5   review_creation_date     99224 non-null  datetime64[us]
 6   review_answer_timestamp  99224 non-null  datetime64[us]
dtypes: datetime64[us](2), int64(1), str(4)
memory usage: 5.3 MB


## 4. Core data-quality checks

Check null values, full-row duplicates, distinct identifiers, and the review-score distribution.

In [10]:
reviews.isna().sum()

review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

In [11]:
reviews.duplicated().sum()

np.int64(0)

In [12]:
reviews["review_id"].nunique()

98410

In [13]:
reviews["order_id"].nunique()

98673

In [14]:
reviews["review_score"].value_counts().sort_index()

review_score
1    11424
2     3151
3     8179
4    19142
5    57328
Name: count, dtype: int64

### Quality findings

- No nulls were found in `review_id`, `order_id`, `review_score`, or the date fields
- `review_comment_title`: **87,656 nulls**
- `review_comment_message`: **58,247 nulls**
- Full duplicate rows: **0**
- Unique `review_id`: **98,410**
- Unique `order_id`: **98,673**
- Review scores contain only the expected values **1–5**

Missing written comments are retained because a review score can still be valid without a text comment.

## 5. Check text lengths

These checks were especially useful because the earlier DBeaver CSV import reported “Data too long” errors.

In [15]:
reviews["review_comment_title"].str.len().max()

np.float64(26.0)

In [16]:
reviews["review_comment_message"].str.len().max()

np.float64(208.0)

### Text-length finding

- Maximum review title length: **26 characters**
- Maximum review message length: **208 characters**

These values fit within the column sizes previously used in MySQL. This confirmed that the DBeaver errors were caused by CSV parsing behavior rather than genuinely oversized review text.

## 6. Investigate identifier uniqueness and table grain

The row count is larger than the number of unique `review_id` and `order_id` values, even though there are no full duplicate rows. The next checks investigate how these identifiers repeat.

In [17]:
review_id_counts = reviews["review_id"].value_counts()

(review_id_counts > 1).sum()

np.int64(789)

In [19]:
review_id_counts.max()

np.int64(3)

In [20]:
reviews[
    reviews["review_id"].duplicated(keep=False)
].sort_values("review_id").head(20)

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
46678,00130cbe1f9d422698c812ed8ded1919,dfcdfc43867d1c1381bfaf62d6b9c195,1,NaN,"O cartucho ""original HP"" 60XL não é reconhecid...",2018-03-07,2018-03-20 18:08:23
29841,00130cbe1f9d422698c812ed8ded1919,04a28263e085d399c97ae49e0b477efa,1,NaN,"O cartucho ""original HP"" 60XL não é reconhecid...",2018-03-07,2018-03-20 18:08:23
90677,0115633a9c298b6a98bcbe4eee75345f,78a4201f58af3463bdab842eea4bc801,5,NaN,NaN,2017-09-21,2017-09-26 03:27:47
63193,0115633a9c298b6a98bcbe4eee75345f,0c9850b2c179c1ef60d2855e2751d1fa,5,NaN,NaN,2017-09-21,2017-09-26 03:27:47
92876,0174caf0ee5964646040cd94e15ac95e,f93a732712407c02dce5dd5088d0f47b,1,NaN,Produto entregue dentro de embalagem do fornec...,2018-03-07,2018-03-08 03:00:53
57280,0174caf0ee5964646040cd94e15ac95e,74db91e33b4e1fd865356c89a61abf1f,1,NaN,Produto entregue dentro de embalagem do fornec...,2018-03-07,2018-03-08 03:00:53
54832,017808d29fd1f942d97e50184dfb4c13,8daaa9e99d60fbba579cc1c3e3bfae01,5,NaN,NaN,2018-03-02,2018-03-05 01:43:30
99167,017808d29fd1f942d97e50184dfb4c13,b1461c8882153b5fe68307c46a506e39,5,NaN,NaN,2018-03-02,2018-03-05 01:43:30
20621,0254bd905dc677a6078990aad3331a36,5bf226cf882c5bf4247f89a97c86f273,1,NaN,O pedido consta de 2 produtos e até agora rece...,2017-09-09,2017-09-13 09:52:44
96080,0254bd905dc677a6078990aad3331a36,331b367bdd766f3d1cf518777317b5d9,1,NaN,O pedido consta de 2 produtos e até agora rece...,2017-09-09,2017-09-13 09:52:44


`duplicated(keep=False)` marks **all rows belonging to a repeated value** as `True`, including the first occurrence. This makes it useful for viewing every row in a repeated group side by side.

In [21]:
reviews.duplicated(
    subset=["review_id", "order_id"]
).sum()

np.int64(0)

In [22]:
order_id_counts = reviews["order_id"].value_counts()

(order_id_counts > 1).sum()

np.int64(547)

In [23]:
order_id_counts.max()

np.int64(3)

In [25]:
reviews[
    reviews["order_id"].duplicated(keep=False)
].sort_values("order_id").head(5)

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
25612,89a02c45c340aeeb1354a24e7d4b2c1e,0035246a40f520710769010f752e7507,5,NaN,NaN,2017-08-29,2017-08-30 01:59:12
22423,2a74b0559eb58fc1ff842ecc999594cb,0035246a40f520710769010f752e7507,5,NaN,Estou acostumada a comprar produtos pelo barat...,2017-08-25,2017-08-29 21:45:57
22779,ab30810c29da5da8045216f0f62652a2,013056cfe49763c6f66bda03396c5ee3,5,NaN,NaN,2018-02-22,2018-02-23 12:12:30
68633,73413b847f63e02bc752b364f6d05ee9,013056cfe49763c6f66bda03396c5ee3,4,NaN,NaN,2018-03-04,2018-03-05 17:02:00
854,830636803620cdf8b6ffaf1b2f6e92b2,0176a6846bcb3b0d3aa3116a9a768597,5,NaN,NaN,2017-12-30,2018-01-02 10:54:06


### Grain and key findings

- **789** different `review_id` values appear more than once
- A `review_id` appears at most **3** times
- **547** different `order_id` values appear more than once
- An `order_id` appears at most **3** times
- Duplicate `(review_id, order_id)` combinations: **0**

Therefore neither `review_id` nor `order_id` is individually unique in this dataset, while the pair `(review_id, order_id)` is unique in the checked data.

**Working grain:** one row represents one **review–order record**.

For future **order-level** analysis, if a single review score is required, the working decision is to use the most recent review based on `review_answer_timestamp`. The raw table itself remains unchanged. The dataset does not prove why multiple review records exist for the same order, so this is treated as observed dataset behavior rather than assuming that a review was edited.

## 7. Load the validated raw data into MySQL

DBeaver's CSV importer had difficulty parsing the review file, so Pandas + SQLAlchemy are used instead.

If the packages are not installed:

```bash
pip install sqlalchemy pymysql
```

`getpass()` keeps the MySQL password out of the notebook source, and `utf8mb4` supports the review text safely.

In [27]:
from sqlalchemy import create_engine, text
from getpass import getpass
from urllib.parse import quote_plus

mysql_user = "root"
mysql_password = getpass("MySQL password: ")
mysql_password = quote_plus(mysql_password)

engine = create_engine(
    f"mysql+pymysql://{mysql_user}:{mysql_password}@localhost:3306/olist_ecommerce?charset=utf8mb4"
)

### Test the connection

Confirm that the SQLAlchemy engine is connected to the intended database.

In [28]:
with engine.connect() as connection:
    database = connection.execute(text("SELECT DATABASE();")).scalar()
    print(database)

olist_ecommerce


### Clear the target table before reload

The earlier failed import may have committed partial batches. `TRUNCATE` clears existing rows while keeping the table structure.

In [29]:
with engine.begin() as connection:
    connection.execute(text("TRUNCATE TABLE order_reviews;"))

### Append the DataFrame to the existing table

Important parameters:

- `if_exists="append"`: keep the existing MySQL table and append rows
- `index=False`: do not create a column from the Pandas row index
- `chunksize=1000`: insert the data in manageable batches
- `method="multi"`: send multiple rows per INSERT for better performance

In [30]:
reviews.to_sql(
    name="order_reviews",
    con=engine,
    if_exists="append",
    index=False,
    chunksize=1000,
    method="multi"
)

99224

## 8. Validate the MySQL load

The final query compares the MySQL row and distinct-ID counts with the Pandas results.

In [31]:
pd.read_sql(
    """
    SELECT
        COUNT(*) AS total_rows,
        COUNT(DISTINCT review_id) AS unique_review_ids,
        COUNT(DISTINCT order_id) AS unique_order_ids
    FROM order_reviews;
    """,
    con=engine
)

,total_rows,unique_review_ids,unique_order_ids
0,99224,98410,98673


## Final result

The Python → MySQL ingestion completed successfully:

| Check | Result |
|---|---:|
| Total rows | **99,224** |
| Unique review IDs | **98,410** |
| Unique order IDs | **98,673** |

The MySQL counts match the Pandas inspection results, confirming that the full review dataset was loaded without losing records.

### Next project step

Add the `order_reviews` findings to the SQL data-quality checks, then continue the SQL business analysis alongside the Python workflow.